src/goto-instrument/contracts/assigns.h

```diff
@@ -14,61 +14,90 @@ Date: July 2021
 #ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_ASSIGNS_H
 #define CPROVER_GOTO_INSTRUMENT_CONTRACTS_ASSIGNS_H
 
+#include <iostream>
 #include <unordered_set>
 
 #include <goto-programs/goto_model.h>
 
+#include <util/c_types.h>
 #include <util/message.h>
 #include <util/pointer_expr.h>
 
+typedef std::pair<const exprt, const exprt> slicet;
+
 /// \brief A class for representing assigns clauses in code contracts
 class assigns_clauset
 {
 public:
-  /// \brief A class for representing targets for assigns clauses
-  class targett
+  /// \brief A class for representing Conditional Address Ranges
+  class conditional_address_ranget
   {
   public:
-    targett(const assigns_clauset &, const exprt &);
+    conditional_address_ranget(const assigns_clauset &, const exprt &);
+
+    exprt generate_inclusion_check(
+      const conditional_address_ranget &lhs,
+      bool skip_lhs_validity_check = false) const;
 
-    static exprt normalize(const exprt &);
+    goto_programt generate_snapshot_instructions() const;
```

Comment:
I added some comments within the functions, but since are short on time, I think we should defer doxygen comments on all of the contracts functions to a separate PR.

Reply:
Agreed.